per-Tensor `grad_fn` for in-place foreach functions

crcrpar opened the pull request:
Generate a `grad_fn` for each (tuple of) `Tensor`(s) of the same index for `_foreach_foo_` and each `grad_fn` is `FooBackward`.

The current status of foreach functions' backward support for the record:
- out-place: Implemented, but no optimized implementations like their forward path
- in-place: not implemented. I think this check https://github.com/pytorch/pytorch/blob/7eaaefafb3ce0e4a0a9e1eb647e340711973ec12/torchgen/api/autograd.py#L309-L311 is partly responsible but the difference of signature between out-place and in-place (see https://github.com/pytorch/pytorch/pull/96405#discussion_r1154690940) would prevent in-place from using out-place versions (the logic is around https://github.com/pytorch/pytorch/blob/7eaaefafb3ce0e4a0a9e1eb647e340711973ec12/torchgen/api/autograd.py#L495-L500)

```c++
void _foreach_abs_(c10::DispatchKeySet ks, at::TensorList self) {
  auto self_ = unpack(self, "self", 0);
  #ifndef NDEBUG
  std::vector<c10::optional<Storage>> self__storage_saved(self_.size());
  for (const Tensor& tensor : self_)
    self__storage_saved.push_back(
      tensor.has_storage() ? c10::optional<Storage>(tensor.storage()) : c10::nullopt);
  std::vector<c10::intrusive_ptr<TensorImpl>> self__impl_saved(self_.size());
  for (size_t i=0; i<self_.size(); i++)
    if (self_[i].defined()) self__impl_saved[i] = self_[i].getIntrusivePtr();
  #endif
  {
    at::AutoDispatchBelowAutograd guard;
    at::redispatch::_foreach_abs_(ks & c10::after_autograd_keyset, self_);
  }
  #ifndef NDEBUG
  for (size_t i=0; i<self_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (self__storage_saved[i].has_value() && !at::impl::tensorlist_has_dispatch(self_))
      AT_ASSERT(self__storage_saved[i].value().is_alias_of(self_[i].storage()));
  }
  for (size_t i=0; i<self_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (self__impl_saved[i] && !at::impl::tensorlist_has_dispatch(self_))
      AT_ASSERT(self__impl_saved[i] == self_[i].getIntrusivePtr());
  }
  #endif
}
```



Related:
- #95431
- #95765 for multiple `grad_fn`s logic

---

Examples: outputs of `_foreach_add_.List`, `_foreach_addcmul_.ScalarList`, and `_foreach_exp`

```c++
void _foreach_addcmul__ScalarList(c10::DispatchKeySet ks, at::TensorList self, at::TensorList tensor1, at::TensorList tensor2, at::ArrayRef<at::Scalar> scalars) {
  auto self_ = unpack(self, "self", 0);
  auto tensor1_ = unpack(tensor1, "tensor1", 1);
  auto tensor2_ = unpack(tensor2, "tensor2", 2);
  auto _any_requires_grad = compute_requires_grad( self, tensor1, tensor2 );

  (void)_any_requires_grad;
  std::vector<c10::optional<at::Tensor>> original_selfs(self.size());
  std::vector<std::shared_ptr<AddcmulBackward0>> grad_fns;
  if (_any_requires_grad) {
    for (const auto& i : c10::irange( self.size() )) {
      const auto ith_requires_grad = compute_requires_grad(self[i], tensor1[i], tensor2[i]);
      check_inplace(self[i], ith_requires_grad);
      grad_fns.push_back([&]() -> std::shared_ptr<AddcmulBackward0> {
          if (!ith_requires_grad) {
              return nullptr;
          } else {
              auto grad_fn = std::shared_ptr<AddcmulBackward0>(new AddcmulBackward0(), deleteNode);
              grad_fn->set_next_edges(collect_next_edges( self[i], tensor1[i], tensor2[i] ));
              return grad_fn;
          }
      }());
    }
    if (!grad_fns.empty()) {

        for (const auto& i : c10::irange(grad_fns.size())) {
            auto grad_fn = grad_fns[i];
            if (grad_fn != nullptr) {
                grad_fn->self_scalar_type = self[i].scalar_type();
                grad_fn->tensor1_scalar_type = tensor1[i].scalar_type();
                if (grad_fn->should_compute_output(1)) {
                  grad_fn->tensor2_ = SavedVariable(tensor2[i], false);
                }
                grad_fn->value = scalars[i];
                if (grad_fn->should_compute_output(2)) {
                  grad_fn->tensor1_ = SavedVariable(tensor1[i], false);
                }
                grad_fn->tensor2_scalar_type = tensor2[i].scalar_type();
            }
        }
    }
  }
  #ifndef NDEBUG
  std::vector<c10::optional<Storage>> self__storage_saved(self_.size());
  for (const Tensor& tensor : self_)
    self__storage_saved.push_back(
      tensor.has_storage() ? c10::optional<Storage>(tensor.storage()) : c10::nullopt);
  std::vector<c10::intrusive_ptr<TensorImpl>> self__impl_saved(self_.size());
  for (size_t i=0; i<self_.size(); i++)
    if (self_[i].defined()) self__impl_saved[i] = self_[i].getIntrusivePtr();
  std::vector<c10::optional<Storage>> tensor1__storage_saved(tensor1_.size());
  for (const Tensor& tensor : tensor1_)
    tensor1__storage_saved.push_back(
      tensor.has_storage() ? c10::optional<Storage>(tensor.storage()) : c10::nullopt);
  std::vector<c10::intrusive_ptr<TensorImpl>> tensor1__impl_saved(tensor1_.size());
  for (size_t i=0; i<tensor1_.size(); i++)
    if (tensor1_[i].defined()) tensor1__impl_saved[i] = tensor1_[i].getIntrusivePtr();
  std::vector<c10::optional<Storage>> tensor2__storage_saved(tensor2_.size());
  for (const Tensor& tensor : tensor2_)
    tensor2__storage_saved.push_back(
      tensor.has_storage() ? c10::optional<Storage>(tensor.storage()) : c10::nullopt);
  std::vector<c10::intrusive_ptr<TensorImpl>> tensor2__impl_saved(tensor2_.size());
  for (size_t i=0; i<tensor2_.size(); i++)
    if (tensor2_[i].defined()) tensor2__impl_saved[i] = tensor2_[i].getIntrusivePtr();
  #endif
  {
    at::AutoDispatchBelowAutograd guard;
    at::redispatch::_foreach_addcmul_(ks & c10::after_autograd_keyset, self_, tensor1_, tensor2_, scalars);
  }
  #ifndef NDEBUG
  for (size_t i=0; i<self_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (self__storage_saved[i].has_value() && !at::impl::tensorlist_has_dispatch(self_))
      TORCH_INTERNAL_ASSERT(self__storage_saved[i].value().is_alias_of(self_[i].storage()));
  }
  for (size_t i=0; i<self_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (self__impl_saved[i] && !at::impl::tensorlist_has_dispatch(self_))
      TORCH_INTERNAL_ASSERT(self__impl_saved[i] == self_[i].getIntrusivePtr());
  }
  for (size_t i=0; i<tensor1_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (tensor1__storage_saved[i].has_value() && !at::impl::tensorlist_has_dispatch(tensor1_))
      TORCH_INTERNAL_ASSERT(tensor1__storage_saved[i].value().is_alias_of(tensor1_[i].storage()));
  }
  for (size_t i=0; i<tensor1_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (tensor1__impl_saved[i] && !at::impl::tensorlist_has_dispatch(tensor1_))
      TORCH_INTERNAL_ASSERT(tensor1__impl_saved[i] == tensor1_[i].getIntrusivePtr());
  }
  for (size_t i=0; i<tensor2_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (tensor2__storage_saved[i].has_value() && !at::impl::tensorlist_has_dispatch(tensor2_))
      TORCH_INTERNAL_ASSERT(tensor2__storage_saved[i].value().is_alias_of(tensor2_[i].storage()));
  }
  for (size_t i=0; i<tensor2_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (tensor2__impl_saved[i] && !at::impl::tensorlist_has_dispatch(tensor2_))
      TORCH_INTERNAL_ASSERT(tensor2__impl_saved[i] == tensor2_[i].getIntrusivePtr());
  }
  #endif
  if (!grad_fns.empty()) {
      auto differentiable_outputs = flatten_tensor_args( self );
      TORCH_INTERNAL_ASSERT(differentiable_outputs.size() == grad_fns.size());
      for (const auto& i : c10::irange(grad_fns.size())) {
          auto grad_fn = grad_fns[i];
          if (grad_fn != nullptr) {
              rebase_history(differentiable_outputs[i], grad_fns[i]);
          }
      }
  }
}

```

```c++
void _foreach_add__List(c10::DispatchKeySet ks, at::TensorList self, at::TensorList other, const at::Scalar & alpha) {
  auto self_ = unpack(self, "self", 0);
  auto other_ = unpack(other, "other", 1);
  auto _any_requires_grad = compute_requires_grad( self, other );

  (void)_any_requires_grad;
  std::vector<c10::optional<at::Tensor>> original_selfs(self.size());
  std::vector<std::shared_ptr<AddBackward0>> grad_fns;
  if (_any_requires_grad) {
    for (const auto& i : c10::irange( self.size() )) {
      const auto ith_requires_grad = compute_requires_grad(self[i], other[i]);
      check_inplace(self[i], ith_requires_grad);
      grad_fns.push_back([&]() -> std::shared_ptr<AddBackward0> {
          if (!ith_requires_grad) {
              return nullptr;
          } else {
              auto grad_fn = std::shared_ptr<AddBackward0>(new AddBackward0(), deleteNode);
              grad_fn->set_next_edges(collect_next_edges( self[i], other[i] ));
              return grad_fn;
          }
      }());
    }
    if (!grad_fns.empty()) {

        for (const auto& i : c10::irange(grad_fns.size())) {
            auto grad_fn = grad_fns[i];
            if (grad_fn != nullptr) {
                grad_fn->other_scalar_type = other[i].scalar_type();
                grad_fn->alpha = alpha;
                grad_fn->self_scalar_type = self[i].scalar_type();
            }
        }
    }
  }
  #ifndef NDEBUG
  std::vector<c10::optional<Storage>> self__storage_saved(self_.size());
  for (const Tensor& tensor : self_)
    self__storage_saved.push_back(
      tensor.has_storage() ? c10::optional<Storage>(tensor.storage()) : c10::nullopt);
  std::vector<c10::intrusive_ptr<TensorImpl>> self__impl_saved(self_.size());
  for (size_t i=0; i<self_.size(); i++)
    if (self_[i].defined()) self__impl_saved[i] = self_[i].getIntrusivePtr();
  std::vector<c10::optional<Storage>> other__storage_saved(other_.size());
  for (const Tensor& tensor : other_)
    other__storage_saved.push_back(
      tensor.has_storage() ? c10::optional<Storage>(tensor.storage()) : c10::nullopt);
  std::vector<c10::intrusive_ptr<TensorImpl>> other__impl_saved(other_.size());
  for (size_t i=0; i<other_.size(); i++)
    if (other_[i].defined()) other__impl_saved[i] = other_[i].getIntrusivePtr();
  #endif
  {
    at::AutoDispatchBelowAutograd guard;
    at::redispatch::_foreach_add_(ks & c10::after_autograd_keyset, self_, other_, alpha);
  }
  #ifndef NDEBUG
  for (size_t i=0; i<self_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (self__storage_saved[i].has_value() && !at::impl::tensorlist_has_dispatch(self_))
      TORCH_INTERNAL_ASSERT(self__storage_saved[i].value().is_alias_of(self_[i].storage()));
  }
  for (size_t i=0; i<self_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (self__impl_saved[i] && !at::impl::tensorlist_has_dispatch(self_))
      TORCH_INTERNAL_ASSERT(self__impl_saved[i] == self_[i].getIntrusivePtr());
  }
  for (size_t i=0; i<other_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (other__storage_saved[i].has_value() && !at::impl::tensorlist_has_dispatch(other_))
      TORCH_INTERNAL_ASSERT(other__storage_saved[i].value().is_alias_of(other_[i].storage()));
  }
  for (size_t i=0; i<other_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (other__impl_saved[i] && !at::impl::tensorlist_has_dispatch(other_))
      TORCH_INTERNAL_ASSERT(other__impl_saved[i] == other_[i].getIntrusivePtr());
  }
  #endif
  if (!grad_fns.empty()) {
      auto differentiable_outputs = flatten_tensor_args( self );
      TORCH_INTERNAL_ASSERT(differentiable_outputs.size() == grad_fns.size());
      for (const auto& i : c10::irange(grad_fns.size())) {
          auto grad_fn = grad_fns[i];
          if (grad_fn != nullptr) {
              rebase_history(differentiable_outputs[i], grad_fns[i]);
          }
      }
  }
}


...

void _foreach_exp_(c10::DispatchKeySet ks, at::TensorList self) {
  auto self_ = unpack(self, "self", 0);
  auto _any_requires_grad = compute_requires_grad( self );

  (void)_any_requires_grad;
  std::vector<c10::optional<at::Tensor>> original_selfs(self.size());
  std::vector<std::shared_ptr<ExpBackward0>> grad_fns;
  if (_any_requires_grad) {
    for (const auto& i : c10::irange( self.size() )) {
      const auto ith_requires_grad = compute_requires_grad(self[i]);
      check_inplace(self[i], ith_requires_grad);
      grad_fns.push_back([&]() -> std::shared_ptr<ExpBackward0> {
          if (!ith_requires_grad) {
              return nullptr;
          } else {
              auto grad_fn = std::shared_ptr<ExpBackward0>(new ExpBackward0(), deleteNode);
              grad_fn->set_next_edges(collect_next_edges( self[i] ));
              return grad_fn;
          }
      }());
    }
  }
  #ifndef NDEBUG
  std::vector<c10::optional<Storage>> self__storage_saved(self_.size());
  for (const Tensor& tensor : self_)
    self__storage_saved.push_back(
      tensor.has_storage() ? c10::optional<Storage>(tensor.storage()) : c10::nullopt);
  std::vector<c10::intrusive_ptr<TensorImpl>> self__impl_saved(self_.size());
  for (size_t i=0; i<self_.size(); i++)
    if (self_[i].defined()) self__impl_saved[i] = self_[i].getIntrusivePtr();
  #endif
  {
    at::AutoDispatchBelowAutograd guard;
    at::redispatch::_foreach_exp_(ks & c10::after_autograd_keyset, self_);
  }
  #ifndef NDEBUG
  for (size_t i=0; i<self_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (self__storage_saved[i].has_value() && !at::impl::tensorlist_has_dispatch(self_))
      TORCH_INTERNAL_ASSERT(self__storage_saved[i].value().is_alias_of(self_[i].storage()));
  }
  for (size_t i=0; i<self_.size() && !at::impl::dispatch_mode_enabled(); i++) {
    if (self__impl_saved[i] && !at::impl::tensorlist_has_dispatch(self_))
      TORCH_INTERNAL_ASSERT(self__impl_saved[i] == self_[i].getIntrusivePtr());
  }
  #endif
  if (!grad_fns.empty()) {
      auto differentiable_outputs = flatten_tensor_args( self );
      TORCH_INTERNAL_ASSERT(differentiable_outputs.size() == grad_fns.size());
      for (const auto& i : c10::irange(grad_fns.size())) {
          auto grad_fn = grad_fns[i];
          if (grad_fn != nullptr) {
              rebase_history(differentiable_outputs[i], grad_fns[i]);
          }
      }
  }
  if (!grad_fns.empty()) {

      for (const auto& i : c10::irange(grad_fns.size())) {
          auto grad_fn = grad_fns[i];
          if (grad_fn != nullptr) {
              grad_fn->result_ = SavedVariable(self[i], true, self[i].is_view());
          }
      }
  }
}

```

crcrpar commented:
@pytorchbot rebase

albanD commented:
@soulitzer can you handle this?

soulitzer commented:
Thanks for the addition! Could you add the codegen output to the PR description?

I mainly had a comment about whether we could just reuse the original node, e.g. instead of having InplaceForeachSin, we'd have just ForeachSin

soulitzer commented on torchgen/api/autograd.py:
```diff
@@ -445,8 +465,10 @@ def gen_foreach_derivativeinfo(
+        op="{}Foreach{}{}".format(
+            "" if is_functional else "Inplace",
+            ref_diff_info.op,
+            foreach_function.func.name.overload_name,
```
> IIUC, torch/csrc/autograd/generated/Functions(.h | .cpp) are generated from DifferentiabilityInfo which is a frozen dataclass, thus I found it straightforward to define new ones.

If are reusing the out-of-place backward nodes, you won't need to modify Functions.h/cpp.

> adjust forward derivatives by using the existing function nodes

Yeah, we would need to adjust the forward formulas to be in-place, but there is already code doing that in match_differentiability_infos, and if we look at the data structure is produced, it does not modify DifferentiabilityInfo instead it has this new field that can be specific to the in-place variant.

https://github.com/pytorch/pytorch/blob/b2f1edabfe6c4d0b589ecd45e62e12e362d70208/torchgen/api/autograd.py#L252-L256

soulitzer commented:
Btw it looks like the `InplaceForeachAddBackward0List` node is saving `self` and `other` just so that we can use self.scalar_type() and other.scalar_type() during backward. If you look at `AddBackward0`, it avoids that by saving `self_scalar_type` and `other_scalar_type`. This should already be handled by REPLACEMENTS from load_derivatives.py, so not sure what is going on here (but maybe it will solve itself if we reuse the AddBackward0 node? :P)

I guess I'm just not convinced that we need a separate node class. I think the two nodes classes should do the exact same thing, and all the changes we need, include those that update forward AD should be done in `gen_variable_type`.

crcrpar commented:
the reason not using the existing function structs defined in `Functions.(cpp/h)` is that I thought it would be demanding to appropriately map expr/name of the existing ones to those of foreach functions' signatures. Will try to figure out

crcrpar committed:
Let in-place foreach functions use `DifferentiabilityInfo`

Next: let `emit_body` work

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
manage to create `original_selfs` and `grad_fns`

Next step is to appropriately add indexing of `[i]` and mapping of
argument names

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Correct suffix `[i]` for in-place foreach w/o DifferentiabilityInfo

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
update test

- make sure Tensor._version is bumped after going through in-place
foreach functions, especially their fast path (or multi_tensor_apply)
- make sure each modified tensor has the same grad_fn with different
Tensor, Scalar, and scalar_types among the list

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
make var name self-explanatory

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
remove checks for my debug

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
add missing `$`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
@soulitzer Could you re-review this? I've managed to use existing `FooBackward` for `_foreach_foo_` functions. The mapping of arguments might look clumsy though. I will resolve the conflict once viable/strict gets the commit

soulitzer reviewed (commented):
Thanks for the update! Looks pretty good so far, but one optimization I think is quite important but not handled here yet is improving the case where not all inputs require grad - In that case I'd only want to create grad_fn/save tensors/rebase_history for the inputs that require grad.  (Somewhat related is this issue https://github.com/pytorch/pytorch/issues/97575)

crcrpar commented on tools/autograd/gen_variable_type.py:
```diff
@@ -998,7 +1051,11 @@ def find_args_with_derivatives(
+            (len(differentiable_outputs) > 0)
+            # note(crcrpar): In-place foreach functions are a void function.
+            or is_inplace_foreach
+        )
```
IIUC most in-place functions of single output out-place functions actually return a Tensor for example https://github.com/pytorch/pytorch/blob/87f5e92916b876f91a6d534ea80a404484e2c9ff/aten/src/ATen/native/native_functions.yaml#L2429, which is why they don't bother with the condition of `len(differentiable_outputs) > 0`.

In-place foreach functions's `self` argument is `List[Tensor]` which make themselves void. There's a check for this schema:
https://github.com/pytorch/pytorch/blob/9e3b34775b73241463658469b74bb3a2f1c01e79/torchgen/model.py#L1400-L1417

soulitzer replied:
I see, thanks!

crcrpar committed:
check `is_inplace_foreach` first

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
make sure FunctionSchema is not in `differentiability_infos`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented on tools/autograd/gen_variable_type.py:
```diff
@@ -970,7 +1009,21 @@ def gen_differentiable_input(
+                if arg in inplace_foreacharg2refarg:
+                    mapped_arg = inplace_foreacharg2refarg[arg]
+                    if arg.name != mapped_arg.name:
+                        arguments.append(
```
ah, now I don't see any reasons, thanks

crcrpar committed:
replace argument with new one rather than appending it

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
the failing job seems to be just hitting the time limit of 20 minutes.
I guess https://github.com/pytorch/pytorch/pull/97506 is related

```console
test_jit_fuser_te.py::TestTEFuserDynamic::test_dynamic_shapes Command took >20min, returning 124
FINISHED PRINTING LOG FILE of test_jit_fuser_te 5/5 (/var/lib/jenkins/workspace/test/test-reports/test_jit_fuser_te_qxp6uzhw.log)

test_jit_fuser_te 5/5 failed!
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/run_test.py", line 1398, in <module>
    main()
  File "/var/lib/jenkins/workspace/test/run_test.py", line 1356, in main
    raise RuntimeError(
RuntimeError: test_jit_fuser_te 5/5 failed!

Tip: You can keep running tests even on failure by passing --keep-going to run_test.py.
If running on CI, add the 'keep-going' label to your PR and rerun your jobs.

real	50m26.606s
user	104m44.971s
sys	5m42.376s
```
ref: https://github.com/pytorch/pytorch/actions/runs/4577441185/jobs/8083158420#step:12:7079

crcrpar commented:
@pytorchbot rebase

soulitzer commented on test/test_foreach.py:
```diff
@@ -179,6 +199,17 @@ def test_binary_op(self, device, dtype, op, is_fastpath):
+                    if inplace_op.func not in (torch._foreach_minimum_, torch._foreach_maximum_):
+                        torch.autograd.backward(sum([t.clone() for t in tensors]).sum(), inputs=tensors)
+                        grads = [t.grad for t in tensors]
+                        self.assertEqual(grads, [t.grad for t in ref_tensors])
```
Do we have a test where not all inputs tensors require grad?
- We should make sure that that case does not error and the behavior is as-if we did a for loop, i.e. only the outputs corresponding to inputs that require grad should require grad. 
- We should also make sure that only inputs (or outputs that correspond to inputs) that require grad should be saved for backward.

crcrpar replied:
I haven't added that kind of test. what do you think about the test 4941cd22b297a5e1d81221bccaed93aba4d4d8ec added?

soulitzer replied:
Look ok to me! (if we wanted to be more exhaustive we could test more combinations, but this is also fine) btw, is it named "test_inplace_foreach_multi_grad_fn" instead of something specific like "test_inplace_foreach_preserves_requires_grad" because we plan to add more checks there? (I think that would be a good idea :)

soulitzer reviewed (commented):
Thanks for the update. Looks mostly ok, but I haven't looked at the rest of gen_variable_type in detail yet because I'm still not convinced of the current high-level direction:

I think we should be more fine grained about saving/setting requires grad since we're doing multiple nodes anyway. Sure in theory there's no well-defined semantic for foreach - currently this PR implements a multi-node solution as-if we were doing a single node, and thats kind of OK too, like it's not a correctness issue. But I think we're missing out on a lot of benefit we'd inherently get from multi-node. So since we're adding this now, might as well avoid having to come back later and rewrite it the other way.

crcrpar committed:
Set attributes of `grad_fn` iff `compute_requires_grad`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
rebase history on necessity

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
verify that `grad_fn` is set iff self requires grad

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Handle ones w/ no derivatives

there are some in-place foreach functions that don't have the reference
backward definitions. For this, compute `requires_grad` when
initializing `grad_fn`s and store and reuse it afterwards.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

soulitzer commented:
Some comments on the generated code, thanks for updating that:

1) just a nit: I think we can avoid creating a vector of bools because we can just rely on whether grad_fn is nullptr or not, e.g. if a particular input does not requires_grad, instead of creating a grad_fn for that just make it nullptr.

2) Sorry for the back and forth on this, but I think we are still saving tensor unnecessarily - looking at the derivatives.yaml entry for addcmul:
```
- name: addcmul(Tensor self, Tensor tensor1, Tensor tensor2, *, Scalar value=1) -> Tensor
  self: handle_r_to_c(self.scalar_type(), grad)
  tensor1: handle_r_to_c(tensor1.scalar_type(), grad * (tensor2 * value).conj())
  tensor2: handle_r_to_c(tensor2.scalar_type(), grad * (tensor1 * value).conj())
  result: self_t + maybe_multiply(tensor1_t * tensor2_p, value) + maybe_multiply(tensor2_t * tensor1_p, value)
```
I think we should be careful to check whether tensor1 and tensor2 require grad, e.g. if only tensor1 requires grad, we should only need to save tensor2, and conversely if only tensor2 requires grad, we should only save tensor1. 

3) For foreach_exp_  - the case where we save the outputs the function- looks like we produce an empty loop where we usually save the inputs
```
for (const auto& i : c10::irange(grad_fns.size())) {
      auto grad_fn = grad_fns[i];
      if (requires_grad_fns[i]) {
      }
    }
```
still creating SavedVariable and attaching it to grad_fn should be avoided (technically this doesn't matter since we don't rebase history anymore on that), but it can confuse future readers. (relying on grad_fn to be nullptr for some inputs would force this to be the case)

crcrpar committed:
set `grad_fn` to `nullptr` if not grads required

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Enable `guard_for` for in-place foreach functions

in order to avoid redundant storing of tensors

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
rename to `test_inplace_foreach_preserves_requires_grad`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
Thank you so much for the comments :)

> 1. just a nit: I think we can avoid creating a vector of bools because we can just rely on whether grad_fn is nullptr or not, e.g. if a particular input does not requires_grad, instead of creating a grad_fn for that just make it nullptr.

e927084 for this

> 2\. Sorry for the back and forth on this, but I think we are still saving tensor unnecessarily - looking at the derivatives.yaml entry for addcmul:

I hadn't paid enough attention to `guard_for` in `emit_save_inputs`. ef05155 should do the work for it🤞 

> 3\. For foreach_exp_  - the case where we save the outputs the function- looks like we produce an empty loop where we usually save the inputs

this should also be addressed in e927084fe1528c83e18a4766e4ef084adee824e1

soulitzer commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8229,12 +8230,13 @@ def _should_disable_fastpath(self, opinfo, rightmost_arg, rightmost_arg_type, dt
 
         # zero_size tensor
-        if dtype == torch.float32 and ("cuda" in device):
+        if dtype == torch.float32 and (not num_input_tensors_specified) and ("cuda" in device):
```
why are we excluding this case? (just curious)

crcrpar replied:
To skip inputs of a bunch of zero-size tensors. See https://github.com/pytorch/pytorch/pull/95028 for the details.

soulitzer commented on test/test_foreach.py:
```diff
@@ -384,9 +428,47 @@ def test_unary_op(self, device, dtype, op, is_fastpath):
+                self.assertEqual(hook_buffer, list(range(num_tensors - 1, -1, -1)))
+
+                ref_inplace_input_tensors = [t.clone().detach().requires_grad_() for t in inplace_input_tensors]
+                ref_inplace_inputs = [0.1 + t for t in ref_inplace_input_tensors]
```
nit: do clone instead of 0.1 + t just for consistency as its usually the op we use when we want to avoid operating on a leaf.

soulitzer reviewed (commented):
Thanks for the update, codegen'd code looks great! 

Some comments on testing
- Since a lot of the logic between the tests feels quite similar maybe its worth factoring that out into a separate test.
- Some other small tests that we probably want:
  - only the necessary tensors are saved if not all inputs require grad (this can probably be combined into that other test)
   - inplace over view should trigger the rebasing logic, i.e. copyslices/as_strided nodes are created
   - inplace on a leaf tensor should error

soulitzer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -948,6 +985,17 @@ def emit_body(
+            foreach_arg_type = getattr(foreach_arg.type, "elem", foreach_arg.type)
+            assert foreach_arg_type == ref_arg.type
+            inplace_foreacharg2refarg[foreach_arg] = ref_arg
+
```
I think its better to replace`inplace_foreacharg2refarg` with `is_inplace_foreach` and then doing `assert info is not None` inside instead (if possible, if there some problematic ops I'm curious what those are). The info is not None case should be handled elsewhere.

crcrpar replied:
hmm, I thought it was enough because a container of 0 elements (here `inplace_foreach2refarg` is a dictionary) is evaluated as `False` in Python. I do know I'm biased here as I've played with this thing. I'll follow your suggestion

crcrpar commented on tools/autograd/gen_variable_type.py:
```diff
@@ -1483,20 +1629,52 @@ def emit_save_outputs() -> str:
+                    f"_any_requires_grad |= compute_requires_grad( {arg} );"
+                    for arg in args
+                ]
+                + [extra_condition, "(void)_any_requires_grad;"]
```
I tried that and it didn't work at that time, didn't look into why. Will give another shot

crcrpar commented on torchgen/api/autograd.py:
```diff
@@ -359,7 +363,11 @@ def gen_foreach_derivativeinfo(
+
+    # non out-place uses the existing Derivative.
+    if foreach_function.func.kind() != SchemaKind.functional:
+        return ref_diff_info, False
```
I thought there's foreach functions with modifiable `out` argument with `SchemaKind.out`. I haven't checked if they have appropriate backward or not

soulitzer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -948,6 +985,17 @@ def emit_body(
+    inplace_foreacharg2refarg: Dict[Argument, Argument] = {}
+    if is_inplace_foreach and info is not None:
+        for foreach_arg, ref_arg in zip(
+            f.func.arguments.flat_non_out, info.func.func.arguments.flat_non_out
```
hmm if not, we should probably leave a comment about why its okay to ignore the difference

soulitzer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -606,6 +614,22 @@
+    """\
+for (const auto& i : c10::irange( ${irange} )) {
+  grad_fns.push_back([&]() -> std::shared_ptr<${op}> {
+      if (!compute_requires_grad(${args_with_derivatives})) {
```
good point, whoops

crcrpar committed:
fix docstring

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
better check: ListType

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
remove redundant logical OR

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
make it todo to support foreach forward AD

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
indent for clarity

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
todo: comment

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
cosmetic change to reduce one nest of if-else

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
remove dead variable

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
code template for `grad_fns`

to refactor save_inputs and save_outputs

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
conservatively checking size of vectors

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
lintrunner

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
clarify the difference between in-place foreach and their counterparts

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
clean up by utilizing `_compute_requires_grad` handles multiple TensorLists

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
cosmetic changes around `SETUP_ANY_REQUIRES_GRAD`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
hopefully clearer condition to use `inplace_foreacharg2refarg`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
why compute_requires_grad again?

the compute_requires_grad in the code template could look completely
nonsense at glance but it actually does have its role.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
rewrite with better code template

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
YAGNI epilog for now

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
rationally use `InplaceForeachVersionBumpCheck` in binary op test

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
ameliorate test_binary_op

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
add inplace & leaf check

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
add inplace & leaf check

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Revert "add inplace & leaf check"

This reverts commit ae468e9ce89ea8027e8086084a54b5200937bf5e.
Seems like the check unexpectedly raises in
`test_inplace_foreach_perserves_requires_grad`

crcrpar committed:
better tensor prep

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
clone detach the original ones, rather than create new ones

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
better test_lerp

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
lintrunner

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented on test/test_foreach.py:
```diff
@@ -384,9 +428,47 @@ def test_unary_op(self, device, dtype, op, is_fastpath):
                 ]
                 ref_tensors = [t.clone().detach().requires_grad_() for t in tensors]
-                sum(wrapped_op.func(tensors)).mean().backward()
+                out = wrapped_op.func(tensors)
```
addressed

crcrpar commented on test/test_foreach.py:
```diff
@@ -384,9 +428,47 @@ def test_unary_op(self, device, dtype, op, is_fastpath):
+                for i, t in enumerate(inplace_inputs):
+                    c = t.clone()
+                    c.register_hook(get_hook(i))
+                    cloned_tensors.append(c)
```
excuse me I don't quite understand "to register the hook to the grad_fn".
could you elaborate a bit?

soulitzer replied:
Ahh yeah this is more of an obscure feature, you can find more information in these docs that were recently added https://pytorch.org/docs/stable/generated/torch.autograd.graph.Node.register_hook.html#torch.autograd.graph.Node.register_hook

crcrpar reviewed (commented):
updated the example codegen outputs in the description

crcrpar committed:
test with backward hook

see:
https://pytorch.org/docs/stable/generated/torch.autograd.graph.Node.register_hook.html

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

soulitzer reviewed (commented):
Thanks for the update, this looks quite good. I think its almost ready to go, just some final comments on code organization in some places and a small issue on testing.

crcrpar committed:
use TORCH_CHECK rather than TORCH_INTERNAL_ASSERT

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
assert output name for inplace

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
dissect into const sets for clarity

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
reverse map for clarity

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
jotting down todos

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
comments, clone, skip

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
fix

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented on test/test_foreach.py:
```diff
@@ -179,6 +202,22 @@ def test_binary_op(self, device, dtype, op, is_fastpath):
+                        inplace_ref([ref_tensors, rhs_arg])
+                        torch.autograd.backward(sum([t.clone() for t in tensors]).sum(), inputs=tensors)
+                        torch.autograd.backward(sum([t.clone() for t in ref_tensors]).sum(), inputs=ref_tensors)
+                        self.assertEqual([t.grad for t in tensors], [t.grad for t in ref_tensors])
```
ah yes as they would attach `NotImplemented` as grad_fn

crcrpar committed:
check schema is inplace

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
@pytorchbot rebase

soulitzer commented on test/test_foreach.py:
```diff
@@ -179,6 +202,22 @@ def test_binary_op(self, device, dtype, op, is_fastpath):
+                    inplace_op([tensors, rhs_arg], is_cuda=False, is_fastpath=False, zero_size=zero_size)
+                    assert_multiple_grad_fns(tensors, self)
+
+                    is_foreach_max_min_imum_with_scalar_or_scalarlist = (
```
Have a comment about why we don't compare with reference for this op

soulitzer commented on test/test_foreach.py:
```diff
@@ -217,6 +256,10 @@ def test_pointwise_op(self, device, dtype, op, is_fastpath):
             self._pointwise_test(
                 wrapped_op, ref, inputs, is_fastpath and not disable_fastpath, False, values=values, zero_size=zero_size
             )
+            with InplaceForeachVersionBumpCheck(self, sample.input):
```
do we want to do the same thing we do for binary test here? e.g. move the context manager to be closer to where we actually call the op

soulitzer commented on test/test_foreach.py:
```diff
@@ -749,6 +856,30 @@ def test_foreach_reduce_large_input(self, device, dtype, op):
+    def test_inplace_foreach_leaf_check_and_grad_fn(self, device, dtype, op):
+        inplace_op = op.inplace_variant
+        if inplace_op is None:
+            self.skipTest("no in-place op available")
```
Why I don't see this skip for any of the tests above?

soulitzer reviewed (approved):
Thanks, good work! 

For some of the todos, do we want to have a central place to track them to do later, or do you plan to investigate them some more before landing?

crcrpar commented on tools/autograd/gen_variable_type.py:
```diff
@@ -966,7 +1055,19 @@ def gen_differentiable_input(
+            for i, arg in enumerate(f.func.arguments.flat_non_out):
+                if arg in inplace_foreacharg2refarg:
+                    mapped_arg = inplace_foreacharg2refarg[arg]
+                    if arg.name != mapped_arg.name:
```
> we could still replace if they are the same

I'd say it's possible.

> we could assert that if the name is the same the Argument is exactly the same as well?

Would this "assert" mean to make sure that "type" is the same if "name" is the same?
Then I don't think so. The major example is self argument. Foreach functions' self is a TensorList but most others' self is a Tensor

soulitzer replied:
Hmm so in that case, why are replacing and not replacing both okay if the Arguments are different?

crcrpar replied:
I thought it was a matter of Argument.name and its type wouldn't mean a lot at least for the following `gen_differentiable_input` call. Either way it's confusing though I removed the if here.

crcrpar commented on tools/autograd/gen_variable_type.py:
```diff
@@ -1232,6 +1406,21 @@ def save_variables(
                 expr = expr + ".vec()"
+
+            # todo(crcrpar): ideally do this replacement before refining `expr` in if-else above.
+            # Naively moving this processing to before if-else seems to cause some weirdos.
```
I believe I misunderstand or overlook something but I saw invalid SavedVariable constructor call with TensorList by moving this block before if-else block.

crcrpar replied:
me neither. I haven't gotten a clear picture here.

crcrpar committed:
more comment

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
> do we want to have a central place to track them to do later, or do you plan to investigate them some more before landing?

I'll keep working on this space since I'm certain that there should be some fallouts / undefined details.

crcrpar committed:
leave a comment

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
narrow the scope of `InplaceForeachVersionBumpCheck`

also, solve redundant inplace ops test from pointwise

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

soulitzer commented:
> I'll keep working on this space since I'm certain that there should be some fallouts / undefined details.

Ok, let me know when you'd like me to take another look!

crcrpar commented:
oh excuse me if if it was unclear, I meant I will keep working on this foreach & torchgen thing in follow-up PRs as a long-running big PR is harder (thank you very much for your patience)

@pytorchbot label ciflow/trunk

crcrpar committed:
replace Argument with correct name and type

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

soulitzer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -1060,13 +1060,12 @@ def gen_differentiable_inputs(f: NativeFunction) -> List[DifferentiableInput]:
-                            mapped_arg.default,
-                            mapped_arg.annotation,
-                        )
+                    arguments[i] = Argument(
```
Add a comment that name is the only thing that matters?

crcrpar committed:
leave a comment

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
@pytorchbot merge
